packages/block-library/src/search/edit.js

```diff
@@ -4,40 +4,46 @@
 import { RichText } from '@wordpress/editor';
 import { __ } from '@wordpress/i18n';
 
-export default function SearchEdit( { className, attributes, setAttributes } ) {
-	const { label, placeholder, buttonText } = attributes;
+let SearchEdit;
 
-	return (
-		<div className={ className }>
-			<RichText
-				wrapperClassName="wp-block-search__label"
-				aria-label={ __( 'Label text' ) }
-				placeholder={ __( 'Add label…' ) }
-				keepPlaceholderOnFocus
-				formattingControls={ [] }
-				value={ label }
-				onChange={ ( html ) => setAttributes( { label: html } ) }
-			/>
-			<input
-				className="wp-block-search__input"
-				aria-label={ __( 'Optional placeholder text' ) }
-				// We hide the placeholder field's placeholder when there is a value. This
-				// stops screen readers from reading the placeholder field's placeholder
-				// which is confusing.
-				placeholder={ placeholder ? undefined : __( 'Optional placeholder…' ) }
-				value={ placeholder }
-				onChange={ ( event ) => setAttributes( { placeholder: event.target.value } ) }
-			/>
-			<RichText
-				wrapperClassName="wp-block-search__button"
-				className="wp-block-search__button-rich-text"
-				aria-label={ __( 'Button text' ) }
-				placeholder={ __( 'Add button text…' ) }
-				keepPlaceholderOnFocus
-				formattingControls={ [] }
-				value={ buttonText }
-				onChange={ ( html ) => setAttributes( { buttonText: html } ) }
-			/>
-		</div>
-	);
+if ( process.env.GUTENBERG_PHASE === 2 ) {
```

Comment:
Why do we need this here? Isn't the check in the block-library sufficient?

Reply:
The ternary in `block-library/index.js` is enough to make sure the block isn't registed, the other if statements make sure the code is removed from the bundle if the phase isn't `2`.

Reply:
To add to that, my understanding is the other code (search/edit, search/index) would be eliminated automatically if we started using the `sideEffects` property in the package.json of our packages:
https://webpack.js.org/guides/tree-shaking/

> The new webpack 4 release expands on this capability with a way to provide hints to the compiler via the "sideEffects" package.json property to denote which files in your project are "pure" and therefore safe to prune if unused.

I suppose it's a question of how far we want to go. If simply removing code from the public api is the goal, then it becomes a lot simpler to implement. 

If the code needs to be eliminated entirely from the bundle, then the extra logic is needed in each file or we can try using `sideEffects`.

Reply:
@blowery or @jsnajdr, I remember you talking about it in one of the PRs. What do you recommend?

Reply:
I tried with `sideEffects` disabled and it works great. I can't find any JS code inside bundled production file.

Reply:
Yes, specifying `sideEffects: false` and doing the `process.env.GUTENBERG_PHASE === 2` check only in the block library is the best way forward.

> Isn't the check in the block-library sufficient?

Without `sideEffect: false`, the `search` block would still be imported and bundled. The transformed code would be:
```js
import * as search from './search';

registerBlocks( [ false ? search : undefined ] );
```

The fact that the `search` import is not used is not a sufficient reason to eliminate it. The `./search` module's code can have side effects (e.g., `window.wp.search = { ... }`) that need to be executed. The `sideEffects: false` flag is needed to tell webpack that the elimination is safe and desired.

By the way, the way how each block's `name` and `settings` are exported as distinct named exports and then treated as one object (`import * as search`) was surprising to see. It would be more natural to treat the whole `{ name, settings }` object as one unit and export is as default. Or is there any case where the exports are cherry-picked, like in:
```js
import { name as searchName } from '@wordpress/block-library/search';
```
?

Reply:
> By the way, the way how each block's `name` and `settings` are exported as distinct named exports and then treated as one object (`import * as search`) was surprising to see. It would be more natural to treat the whole `{ name, settings }` object as one unit and export is as default. Or is there any case where the exports are cherry-picked, like in:
> 
> ```js
> import { name as searchName } from '@wordpress/block-library/search';
> ```
> 
> ?

Agreed, I think I coded it this way a long time ago and I don't remember why :) We are going to change it in the near future as we are finalizing how blocks can be registered on the server and client using shared settings with JSON definition. See #13693. There are going to be new challenges :)

Reply:
Many thanks for explaining how Webpack uses `sideEffects` to eliminate dead code. I will investigate which packages can benefit from it. There are lots of packages which don't run any side-effects when they are loaded in the browser. I think the only issue we will have with those which register stores.

Reply:
> I think the only issue we will have with those which register stores.

That's right, we don't need to go far to see a typical module side effect 🙂 The store registrations, namely lack of control over the registration order, caused a few headaches in the Calypso integration project by the way.

Reply:
Yep, every call to registerStore seems to cause side effects. We'd have to list those files or find a way to not cause side-effects.

Block registration seems ok though.

I've made an issue here to track adding the property:
https://github.com/WordPress/gutenberg/issues/13910